[4.0][webservice] API pagination links

stefanoel opened the pull request:
Pull Request for Issue #27004 .

### Summary of Changes
\libraries\src\MVC\View\JsonApiView.php was modified adding `model->setState` because 'start' and 'limit' weren't populated and they always got default values (0 and 20) showing the first 20 records.
I don't know if there's a better solution than this one to achieve the same result...
Furthermore, I've added checks to omit unnecessary pagination links.

com_config APIs now return correct links.

I'm not sure about the best way/place to encode square brackets...

### Testing Instructions
#### \libraries\src\MVC\View\JsonApiView.php
/api/index.php/v1/plugins
/api/index.php/v1/plugins?page[offset]=10&page[limit]=10 _(a generic page)_
/api/index.php/v1/plugins?page[offset]=110&page[limit]=10 _(last page)_

#### \api\components\com_config\View\Application\JsonapiView.php
/api/index.php/v1/config/application

#### \api\components\com_config\View\Component\JsonapiView.php
/api/index.php/v1/config/com_content
/api/index.php/v1/config/com_content?page[offset]=40&page[limit]=20 _(a generic page)_
/api/index.php/v1/config/com_content?page[offset]=60&page[limit]=20 _(last page)_

### Expected result

### Actual result

### Documentation Changes Required

stefanoel committed:
Pagination fix

API com_config now return correct links for pagination.
Similarly, fixed JsonApiView.php.
In JsonApiView I added model->setState because 'start' and 'limit' weren't populated and they always got default values (0 and 20) showing the first 20 records.
I don't know if there's a better solution than this one...

stefanoel committed:
Merge branch '4.0-dev' into ws-pagination

stefanoel committed:
Code standard fixes

stefanoel committed:
Merge branch 'ws-pagination' of https://github.com/stefanoel/joomla-cms-wstest into ws-pagination

brianteeman commented:
From what is described this PR does exactly what it says. I dont know enough to determine if it is correct to remove the  unnecessary pagination links but it does what it says

wilsonge commented on libraries/src/MVC/View/JsonApiView.php:
```diff
@@ -112,6 +112,15 @@ public function displayList(array $items = null)
+		$currentPageQuery = $currentUrl->getVar('page', $currentPageDefaultInformation);
+
+		// Set model start and limit params
+		$model->setState('list.start', $currentPageQuery['offset']);
```
This should proxy through from the code here https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/MVC/Controller/ApiController.php#L214

stefanoel replied:
I'm not sure I understood what you mean...
I moved `$model->setState` from JsonApiView.php to ApiController.php, using values obtained from the input page param (https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/MVC/Controller/ApiController.php#L201), and it seems to work fine.
So, can I commit these changes?
But in that case, I'd like to ask some clarifications on how the pagination params are stored in $this->input
https://github.com/joomla/joomla-cms/blob/8635691c325610cc034304bc17b85ec063b2769f/libraries/src/MVC/Controller/ApiController.php#L202-L214
Here is used an array ($internalPaginationMapping) to store only the limit param, and moreover, dumping $this->input I noticed that the array isn't stored as associative, 'limit' key is dropped. Could these line be reviewed or this way of store pagination params in input are mandatory?

wilsonge commented:
> I dont know enough to determine if it is correct to remove the unnecessary pagination links but it does what it says

It's largely a personal preference kinda thing. JSON API as a spec doesn't have strong opinions https://jsonapi.org/format/#fetching-pagination it just says "Keys MUST either be omitted or have a null value to indicate that a particular link is unavailable.". My feeling is first/last as a link is available so should be present (even if it's the current page). But I would definitely agree that having previous/next as nullable makes sense. As with all these subjective things - it's always down to personal preference :)

brianteeman commented:
>  My feeling is first/last as a link is available so should be present (even if it's the current page)

In the gui they would be disabled

noccioletta commented:
Hi, I'd like to better understand your responses:
is the workaround proposed by @stefanoel  consistent or not ?
Is it worthwhile for Joomla4 ?

wilsonge commented:
The changes to next/previous page is a good change. I *think* the first/last links should always be present.
I also think the view shouldn't be setting state into the model - it should be handled in the controller already.

chrisdavenport commented:
The question of whether to include first/last/next/previous links on pages where they are redundant is one I've thought about before.  As a rule of thumb I prefer to avoid them.  My reasoning is that links are pointers to possible state transitions (that's what HATEOAS is all about), but redundant links are a bit like empty promises; it looks like they should lead to a state transition but actually they just lead straight back to the current state.

The client needs to know definitively that it has reached the first/last page when paginating.  Including redundant links means that the client must compare the current URI with the first/last link URI and decide if they're the same or not.  This can sometimes be tricky if the request URI can include arguments that don't affect the result set: how does the client know to ignore them when URI structure is supposed to be opaque?  A simple string comparison may not suffice; it must reduce both to a canonical form before the comparison.  It's only really safe to include redundant links if there are other signals that the client can use to decide whether it needs to follow a link or not.  For example, if there are separate pageNumber and pageCount attributes returned in the response.

So, in my opinion, if you include redundant links, you should also include pageNumber and pageCount (or something similar) and encourage developers not to do URI comparisons.  On the other hand, omitting redundant links doesn't mean that you should omit pageNumber/pageCount as they can be used by clients that want to randomly access pages.

So my preference is to omit redundant links but include pageNumber/pageCount (or better still something like Offset and totalCount as the concept of a page can be a bit problematic).  And on a first pass at least, assume that the data is essentially static so you don't have to worry about insertions and deletions accidentally resulting in duplications and omissions when paginating; that's a whole other problem to solve.

wilsonge commented:
There already is a total-pages property available. 

> The client needs to know definitively that it has reached the first/last page when paginating

That's why you use the next/previous attributes for in JSON API

> This can sometimes be tricky if the request URI can include arguments that don't affect the result set

These are always going to be query parameters. If the URL path is physically different you cannot guarantee the data is the same (and in our core Joomla response this is the case anyhow for the purposes of said checks - which is the more concrete thing - because ultimately people are parsing our responses).

I don't see first and last pages are not **necessarily** redundant even if they are the same page. Previous/Next definitely would be redundant because it's not actually a next/previous page if they don't exist

stefanoel committed:
Merge branch '4.0-dev' into ws-pagination

stefanoel committed:
Updated ApiController.php displayList() pagination

stefanoel committed:
Update JsonApiView.php

Removed $model->setState from the view, everything seems to work...
ToDo: square brackets encoding?

stefanoel committed:
Square bracket encoding in pagination links

I introduced the method "queryEncode" to encode square brackets in pagination links.
AbstractURI->toString() returns unencoded brackets, so I had to find a workaround, but I'm not sure it's the cleanest way to do this.
The method is defined as protected in Joomla\CMS\MVC\View\JsonApiView so that it can be used by other extending classes, like com_config api views.
Fixed an error in com_config component view pagination (L109).

stefanoel committed:
Merge branch '4.0-dev' into ws-pagination

stefanoel commented:
About square brackets encoding

Pagination links have unencoded square brackets, due to AbstractURI buildQuery method  (https://github.com/joomla-framework/uri/blob/eed89d4997fcc52c08bba56116fa3ac4aef6a9db/src/AbstractUri.php#L345-L348) as it was already reported here

- https://github.com/joomla/joomla-cms/issues/17365
- https://github.com/joomla-framework/uri/pull/21
- https://github.com/joomla/joomla-cms/issues/17966

so, to comply with JSON API spec, I chose to handle the encoding with an ad hoc protected method (I've also updated the com_config APIs to use it)
https://github.com/stefanoel/joomla-cms-wstest/blob/5df64298d7445ba3183923b00e65bcf5df434e1e/libraries/src/MVC/View/JsonApiView.php#L265

Now the links returned have encoded square brackets.
Any suggestions?

stefanoel committed:
Merge branch '4.0-dev' into ws-pagination

stefanoel committed:
Merge branch '4.0-dev' into ws-pagination

Ankit7Das commented:
I wish to work on it

wilsonge commented:
@stefanoel apologies this one dropped off my radar. are you willing to fix the conflicts here so i can do a round of testing and get it in please

stefanoel committed:
Update JsonapiView.php

Quy commented on api/components/com_config/View/Application/JsonapiView.php:
```diff
@@ -106,7 +117,6 @@ public function displayList(array $items = null)
 	protected function prepareItem($item)
 	{
 		$item->id = ExtensionHelper::getExtensionRecord('files_joomla')->extension_id;
```
```suggestion
		$item->id = ExtensionHelper::getExtensionRecord('files_joomla')->extension_id;

```

Quy commented on libraries/src/MVC/View/JsonApiView.php:
```diff
@@ -239,4 +252,19 @@ protected function prepareItem($item)
+	{
+		return str_replace(array('[', ']'), array('%5B', '%5D'), $query);
+	}
+
```
```suggestion
```

stefanoel committed:
Update libraries/src/MVC/View/JsonApiView.php

Co-Authored-By: Quy <quy@fluxbb.org>

stefanoel commented:
sorry, now I'm stuck and I can't undestand the problem... I can't access my workstation and I'm not used to solve things from web interface... can you help please?

wilsonge committed:
Update api/components/com_config/View/Application/JsonapiView.php

Co-Authored-By: Quy <quy@fluxbb.org>

Razzo1987 commented:
I have not tested this item.<br /><br />When I try to apply the patch (Joomla! Patch Tester 3.0.0 beta 4):

Error
The file marked for modification does not exist: api/components/com_config/View/Application/JsonapiView.php
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27005">issues.joomla.org/tracker/joomla-cms/27005</a>.</sub>

alikon commented:
can i suggest you to use git commands to apply this patch
`curl -L https://github.com/joomla/joomla-cms/pull/27005.diff | git apply`

patchtester >  Patch Tester 3.0.0 Beta 3 is still a WIP

anyway this pr needs conflict to be solved

alikon commented:
@stefanoel  can you fix conflict please

stefanoel commented:
> can i suggest you to use git commands to apply this patch
> `curl -L https://github.com/joomla/joomla-cms/pull/27005.diff | git apply`
> 
> patchtester > Patch Tester 3.0.0 Beta 3 is still a WIP
> 
> anyway this pr needs conflict to be solved

Hi @alikon can you suggest how to do it from web interface? I'm working from home due to coronavirus emergency here in Italy, and I have to use a smartphone these days. I can't access my workstation...

richard67 commented:
@wilsonge Could you ckeck the conflict in the ApiController? Can it be that the pagination link is already ok with recently merged PR #27696 ? Or does it still need the change from this PR here in addition? I've stumpled over this when checking the merge conflicts here to 4.0-dev.

richard67 committed:
Merge remote-tracking branch 'upstream/4.0-dev' into 4.0-dev-stefanoel-ws-pagination-1

richard67 committed:
Add back removed line from conflicts solving

richard67 commented:
@stefanoel I've made a pull request in your repository: [https://github.com/stefanoel/joomla-cms-wstest/pull/1](https://github.com/stefanoel/joomla-cms-wstest/pull/1). This will update your PR here so it is up to date with its base branch, and solves the merge conflicts shown here on GitHub.

@wilsonge Please review the changes in `libraries/src/MVC/Controller/ApiController.php` as I mentioned above.

stefanoel committed:
Merge pull request #1 from richard67/4.0-dev-stefanoel-ws-pagination-1

Merge upstream/4.0-dev into ws-pagination

richard67 commented:
@wilsonge @alikon Conflicts solved. Ready for review and testing.

alikon commented:
great team-working
 f***** covid19

richard67 commented:
@wilsonge Drone failing the 2nd time now. No idea if related.

wilsonge committed:
Fix up the code to latest version

wilsonge committed:
More fixes

wilsonge commented:
This looks good to me! Thankyou!
